### ISSUE TITLE
doc_markdown lint triggers erroneously wihtin links

### ISSUE THREAD

**Issue by stouset** [rust-lang/rust-clippy#5306]
```
$ cargo clippy -V
clippy 0.0.212 (329923ed 2020-03-04)
```

When using Markdown to link to methods within the generated documentation, this lint triggers erroneously. For instance:

```rust
/// Call [`new`][Foo::new] to instantiate this struct.
struct Foo {
}
```

In this made-up example, the *target* of the link is `Foo::new` per [intra-rustdoc links](https://github.com/rust-lang/rfcs/pull/1946). Adding backticks around it would break the intra-doc linking.

**Comment by jpospychala** [comment]
If I read [intra-rustdoc links](https://github.com/rust-lang/rfcs/pull/1946) correctly, reference link notation (`[label][link]`) requires a reference definition somewhere else in the document, like:

```rust
/// Call [`new`][fnew] to instantiate this struct.
///
/// [fnew]: Foo::new
struct Foo {
}
```

Then link doesn't trigger `doc_markdown` lint.

Otherwise a proper link format is inline link (`[label](link)`):

```rust
/// Call [`new`](Foo::new) to instantiate this struct.
struct Foo {
}
```

**Comment by stouset** [comment]
The reference definitions are unnecessary for struct and method names. They are created automatically. Try it!

**Comment by stouset** [comment]
Oh no... I just reread the PR and it turns out for over a year I've been doing this wrong. Interestingly it still works (or at least it did back when I first used it).
